### PR TITLE
Add Validate to PredicateReorderingRule

### LINKED ISSUE #320: Make PredicateReorderingRule aware of Validate
For the purpose of reordering predicates, we should consider the MVCC validation step as just another predicate with an estimated cardinality of `input_cardinality * table._approx_invalid_row_count / table.row_count()`.

### PR DESCRIPTION
**Pull request by SvenLehmann**
Fixes #320.

Validate can be seen as Predicate on a MVCC column. Therefore it makes sense to treat it like a Predicate in the optimizer rule.

This PR reverts some of the changes that came with #814, in particular removing approx_invalid_row_count. In order to provide statistics for Validate this information is necessary and has thus been added again.

This PR is currently blocked by #1081, which will be resolved once #1078 is merged.

### CODE REVIEW

File: src/lib/statistics/table_statistics.hpp [hunk 1]
```diff
@@ -52,12 +53,20 @@ class TableStatistics final {
   TableStatistics estimate_disjunction(const TableStatistics& right_table_statistics) const;
   /** @} */
 
+  // Increases the (approximate) count of invalid rows in the table (caused by deletes).
+  void increment_invalid_row_count(uint64_t count);
```
Comment (mrks) [inline src/lib/statistics/table_statistics.hpp]: Isn't "increment" in steps of one?

Comment (SvenLehmann) [reply src/lib/statistics/table_statistics.hpp]: This is how it was called before. Would "increase" be more appropriate in your opinion?

Comment (mrks) [reply src/lib/statistics/table_statistics.hpp]: yes

**Comment by mrks** [comment]
Could someone merge the new master plz?

**Comment by SvenLehmann** [comment]
For the sake of completeness:

```
{
	"mvcc": true,
	"scale": 0.1
}
+-----------+-------------------+------+-----------------+------+--------+-----------------------------------------------+
| Benchmark | prev. iter/s      | runs | new iter/s      | runs | change |               p-value (significant if <0.001) |
+-----------+-------------------+------+-----------------+------+--------+-----------------------------------------------+
| TPC-H 1   | 2.17774224281     | 11   | 2.29936051369   | 12   | +6%    |                   (run time too short) 0.0319 |
| TPC-H 2   | 6.25722503662     | 32   | 41.3539543152   | 207  | +561%  |                   (run time too short) 0.0000 |
| TPC-H 3   | 15.7689065933     | 79   | 16.1384773254   | 81   | +2%    |                   (run time too short) 0.0000 |
| TPC-H 4   | 0.0104126641527   | 1    | 0.270576447248  | 2    | +2499% |    (run time too short) (not enough runs) nan |
| TPC-H 5   | 12.9709968567     | 65   | 12.62159729     | 64   | -3%    |                   (run time too short) 0.1069 |
| TPC-H 6   | 22.6617221832     | 114  | 26.6893634796   | 134  | +18%   |                   (run time too short) 0.0000 |
| TPC-H 7   | 4.58197021484     | 23   | 4.76147031784   | 24   | +4%    |                   (run time too short) 0.0276 |
| TPC-H 8   | 15.0345058441     | 76   | 15.6942892075   | 79   | +4%    |                   (run time too short) 0.1841 |
| TPC-H 9   | 8.50787639618     | 43   | 9.11801815033   | 46   | +7%    |                   (run time too short) 0.0000 |
| TPC-H 10  | 10.7408370972     | 54   | 13.2091112137   | 67   | +23%   |                   (run time too short) 0.0000 |
| TPC-H 11  | 63.2970962524     | 317  | 66.5715484619   | 333  | +5%    |                   (run time too short) 0.0000 |
| TPC-H 12  | 14.9326906204     | 75   | 15.5906906128   | 78   | +4%    |                   (run time too short) 0.0000 |
| TPC-H 13  | 14.9215488434     | 75   | 14.7273645401   | 74   | -1%    |                   (run time too short) 0.0655 |
| TPC-H 14  | 28.4480762482     | 143  | 29.3096542358   | 147  | +3%    |                   (run time too short) 0.0000 |
| TPC-H 15  | 13.5528936386     | 68   | 14.1805505753   | 71   | +5%    |                   (run time too short) 0.0000 |
| TPC-H 16  | 9.98592853546     | 50   | 11.3623981476   | 57   | +14%   |                   (run time too short) 0.0000 |
| TPC-H 17  | 0.109338097274    | 1    | 2.49550533295   | 13   | +2182% |    (run time too short) (not enough runs) nan |
| TPC-H 18  | 0.155690550804    | 1    | 0.16587805748   | 1    | +7%    |    (run time too short) (not enough runs) nan |
| TPC-H 19  | 1.737210989       | 9    | 1.85228908062   | 10   | +7%    | (run time too short) (not enough runs) 0.0009 |
| TPC-H 20  | 0.000468558864668 | 1    | 0.0109410630539 | 1    | +2235% |                         (not enough runs) nan |
| TPC-H 21  | 0.00671993242577  | 1    | 0.166862383485  | 1    | +2383% |    (run time too short) (not enough runs) nan |
| TPC-H 22  | 0.01722211577     | 1    | 0.589913368225  | 3    | +3325% |    (run time too short) (not enough runs) nan |
| average   |                   |      |                 |      | +604%  |                                               |
+-----------+-------------------+------+-----------------+------+--------+-----------------------------------------------+
```

**Comment by mrzzzrm** [comment]
How is this even possible?

**Comment by mrks** [comment]
Well, if we don't have any deleted rows, executing the least selective "scan" first is a bad idea :D

**Comment by SvenLehmann** [comment]
Looking at the query plans for TPC-H 2 it seems like simply by reordering:
For this query, it seems to be the Projection that includes the Subselect (152ms vs. 18ms) 
Unfortunately there are no runtimes for the operators within the subselects. But just from looking at the two subquery plans, it seems like pushing down the Predicate(p_partkey = ps_partkey) below Validate makes the difference here. This Predicate filters roughly 80'000 rows down to 4 (based on LQP estimations). When Validate is executed after this filter, it basically has no costs anymore and the Predicate became faster due to not having to working with ReferenceColumns.

The other queries that received a boost have similar characteristics:
- TPC-H 4: Projection with Subselect (94s vs. 4s)
- TPC-H 17: Projection with Subselect (8s vs. 368ms)
- TPC-H 20: Projection with Subselect (35min vs. 1min) 
- TPC-H 21: Two Projections with Subselect (12s vs. 0.5s and 120s vs. 6s)
- TPC-H 22: Projection with Subselect (59s vs. 1s)

But as @mrks pointed out, these numbers are only that good as long as we don't have any deleted rows.

**Comment by Bouncner** [comment]
“Not any deleted rows”?
Would could be problem with filtering even more tuples out?

**Comment by mrks** [comment]
?

**Comment by SvenLehmann** [comment]
Yeah, you're right. It always makes sense to reorder and filter. My last sentence might be misleading.

**Pull request merged**